src/hotspot/os_cpu/linux_zero/os_linux_zero.cpp

```diff
@@ -87,24 +87,122 @@ char* os::non_memory_address_word() {
+    return (address)uc->uc_mcontext.arm_pc;
+#elif defined(AARCH64)
+    return (address)uc->uc_mcontext.pc;
+#elif defined(PPC)
+    return (address)uc->uc_mcontext.regs->nip;
+#elif defined(RISCV)
+    return (address)uc->uc_mcontext.__gregs[REG_PC];
+#elif defined(S390)
+    return (address)uc->uc_mcontext.psw.addr;
+#else
+    // Non-arch-specific Zero code does not really know the PC.
+    // If possible, add the arch-specific definition in this method.
+    fatal("Cannot handle ucontext_get_pc");
+#endif
+  }
+
+  // Answer the default and hope for the best
+  return nullptr;
 }
 
-void os::Posix::ucontext_set_pc(ucontext_t * uc, address pc) {
+void os::Posix::ucontext_set_pc(ucontext_t* uc, address pc) {
   ShouldNotCallThis();
 }
 
+intptr_t* os::Linux::ucontext_get_sp(const ucontext_t* uc) {
+  if (DecodeErrorContext) {
+#if defined(IA32)
+    return (intptr_t*)uc->uc_mcontext.gregs[REG_UESP];
+#elif defined(AMD64)
+    return (intptr_t*)uc->uc_mcontext.gregs[REG_RSP];
+#elif defined(ARM)
+    return (intptr_t*)uc->uc_mcontext.arm_sp;
+#elif defined(AARCH64)
+    return (intptr_t*)uc->uc_mcontext.sp;
+#elif defined(PPC)
+    return (intptr_t*)uc->uc_mcontext.regs->gpr[1/*REG_SP*/];
+#elif defined(RISCV)
+    return (intptr_t*)uc->uc_mcontext.__gregs[REG_SP];
+#elif defined(S390)
+    return (intptr_t*)uc->uc_mcontext.gregs[15/*REG_SP*/];
+#else
+    // Non-arch-specific Zero code does not really know the SP.
+    // If possible, add the arch-specific definition in this method.
+    fatal("Cannot handle ucontext_get_sp");
+#endif
+  }
+
+  // Answer the default and hope for the best
+  return nullptr;
+}
+
+intptr_t* os::Linux::ucontext_get_fp(const ucontext_t* uc) {
+  if (DecodeErrorContext) {
+#if defined(IA32)
+    return (intptr_t*)uc->uc_mcontext.gregs[REG_EBP];
+#elif defined(AMD64)
+    return (intptr_t*)uc->uc_mcontext.gregs[REG_RBP];
+#elif defined(ARM)
+    return (intptr_t*)uc->uc_mcontext.arm_fp;
+#elif defined(AARCH64)
+    return (intptr_t*)uc->uc_mcontext.regs[29 /* REG_FP */];
+#elif defined(PPC)
+    return nullptr;
+#elif defined(RISCV)
+    return (intptr_t*)uc->uc_mcontext.__gregs[8 /* REG_FP */];
+#elif defined(S390)
+    return nullptr;
+#else
+    // Non-arch-specific Zero code does not really know the FP.
+    // If possible, add the arch-specific definition in this method.
+    fatal("Cannot handle ucontext_get_fp");
+#endif
+  }
+
+  // Answer the default and hope for the best
+  return nullptr;
+}
+
 address os::fetch_frame_from_context(const void* ucVoid,
                                      intptr_t** ret_sp,
                                      intptr_t** ret_fp) {
-  ShouldNotCallThis();
-  return NULL; // silence compile warnings
+  address epc;
+  const ucontext_t* uc = (const ucontext_t*)ucVoid;
+
+  if (uc != NULL) {
+    epc = os::Posix::ucontext_get_pc(uc);
+    if (ret_sp) *ret_sp = (intptr_t *)os::Linux::ucontext_get_sp(uc);
+    if (ret_fp) *ret_fp = (intptr_t *)os::Linux::ucontext_get_fp(uc);
```

Comment:
style nits: curly brackets? Remove space in `intptr_t *`?

Reply:
Right, modified.